File: gate/emit_ref.c
/* The other half of the declaration-drift gate: prints the same
 * reference gate/emit_ref.zig prints, but as a C compiler sees
 * include/csar.h — values from the #defines, layout from
 * offsetof/sizeof. Checks the header as compiled, not as text.
 *
 * Hand-listed, in capi.zig's decl order. A constant missing here or
 * in the header diffs against the Zig side, which reflects over
 * everything; a name missing from the header fails this file's
 * compile. */
#include <stdint.h>
#include <stdio.h>
#include <stddef.h>
#include <string.h>
#include "csar.h"

static void num(const char *name, int64_t v) { printf("%s %lld\n", name, (long long)v); }
static void bits(const char *name, double v) {
    uint64_t b;
    memcpy(&b, &v, 8);
    printf("%s bits:%llx\n", name, (unsigned long long)b);
}

int main(void) {
    num("CSAR_OK", CSAR_OK);
    num("CSAR_INSUFFICIENT_POINTS", CSAR_INSUFFICIENT_POINTS);
    num("CSAR_INVALID_TOLERANCE", CSAR_INVALID_TOLERANCE);
    num("CSAR_COPLANAR_INPUT", CSAR_COPLANAR_INPUT);
    num("CSAR_OUT_OF_MEMORY", CSAR_OUT_OF_MEMORY);
    num("CSAR_INTERNAL", CSAR_INTERNAL);
    num("CSAR_INVALID_METHOD", CSAR_INVALID_METHOD);
    num("CSAR_TOO_MANY_POINTS", CSAR_TOO_MANY_POINTS);
    num("CSAR_WASM_MAX_PTS", CSAR_WASM_MAX_PTS);
    num("CSAR_STATUS_CONVERGED", CSAR_STATUS_CONVERGED);
    num("CSAR_STATUS_INFEASIBLE", CSAR_STATUS_INFEASIBLE);
    num("CSAR_STATUS_DID_NOT_CONVERGE", CSAR_STATUS_DID_NOT_CONVERGE);
    num("CSAR_STATUS_PRECISION_FLOOR", CSAR_STATUS_PRECISION_FLOOR);
    num("CSAR_METHOD_TRUST", CSAR_METHOD_TRUST);
    num("CSAR_METHOD_AUTO", CSAR_METHOD_AUTO);
    num("CSAR_STATUS_NONE", CSAR_STATUS_NONE);
    num("CSAR_METHOD_NONE", CSAR_METHOD_NONE);
    bits("CSAR_DEFAULT_GAP_TOL", CSAR_DEFAULT_GAP_TOL);
    num("CSAR_DEFAULT_N_HULL", CSAR_DEFAULT_N_HULL);
    bits("CSAR_DEFAULT_COPLANARITY_TOL", CSAR_DEFAULT_COPLANARITY_TOL);
    num("CSAR_DEFAULT_MAX_OUTER", CSAR_DEFAULT_MAX_OUTER);

    printf("offsetof q %zu\n", offsetof(csar_result, q));
    printf("offsetof sigma %zu\n", offsetof(csar_result, sigma));
    printf("offsetof gap %zu\n", offsetof(csar_result, gap));
    printf("offsetof gap_floor %zu\n", offsetof(csar_result, gap_floor));
    printf("offsetof residual %zu\n", offsetof(csar_result, residual));
    printf("offsetof status %zu\n", offsetof(csar_result, status));
    printf("offsetof method %zu\n", offsetof(csar_result, method));
    printf("offsetof n_iters %zu\n", offsetof(csar_result, n_iters));
    printf("sizeof csar_result %zu\n", (size_t)sizeof(csar_result));
    return 0;
}

File: gate/emit_ref.zig
//! One half of the declaration-drift gate (`zig build gate`): prints
//! the ABI reference by comptime reflection over capi — every pub
//! `CSAR_*` constant and csar_result's layout — in the exact format
//! gate/emit_ref.c prints from include/csar.h. The build step diffs
//! the two.
//!
//! Reflection, not a hand list, on this side: a new CSAR_* decl in
//! capi.zig shows up here automatically and diffs red until the
//! header (and emit_ref.c's list) carry it too.
//!
//! f64 values print as their bit pattern — exact, and immune to
//! printf-vs-Zig float formatting differences.

const std = @import("std");
const capi = @import("capi");

pub fn main(init: std.process.Init) !void {
    var buf: [4096]u8 = undefined;
    var file_writer = std.Io.File.stdout().writer(init.io, &buf);
    const w = &file_writer.interface;

    inline for (@typeInfo(capi).@"struct".decls) |d| {
        if (comptime std.mem.startsWith(u8, d.name, "CSAR_")) {
            const v = @field(capi, d.name);
            switch (@TypeOf(v)) {
                i32, u32 => try w.print("{s} {d}\n", .{ d.name, v }),
                f64 => try w.print("{s} bits:{x}\n", .{ d.name, @as(u64, @bitCast(v)) }),
                else => @compileError("unhandled CSAR_ decl type: " ++ d.name),
            }
        }
    }
    inline for (@typeInfo(capi.Result).@"struct".fields) |f| {
        try w.print("offsetof {s} {d}\n", .{ f.name, @offsetOf(capi.Result, f.name) });
    }
    try w.print("sizeof csar_result {d}\n", .{@sizeOf(capi.Result)});
    try w.flush();
}
